pymc_experimental/inference/find_map.py

```diff
@@ -0,0 +1,415 @@
+
+from collections.abc import Callable
+from typing import cast
+
+import jax
+import numpy as np
+import pymc as pm
+import pytensor
+import pytensor.tensor as pt
+
+from better_optimize import minimize
+from better_optimize.constants import MINIMIZE_MODE_KWARGS, minimize_method
+from pymc.blocking import DictToArrayBijection, RaveledVars
+from pymc.initial_point import make_initial_point_fn
+from pymc.model.transform.optimization import freeze_dims_and_data
+from pymc.pytensorf import join_nonshared_inputs
+from pymc.util import get_default_varnames
+from pytensor.compile import Function
+from pytensor.tensor import TensorVariable
+from scipy.optimize import OptimizeResult
+
+_log = logging.getLogger(__name__)
+
+
+def set_optimizer_function_defaults(method, use_grad, use_hess, use_hessp):
+    method_info = MINIMIZE_MODE_KWARGS[method].copy()
+
+    use_grad = use_grad if use_grad is not None else method_info["uses_grad"]
+    use_hess = use_hess if use_hess is not None else method_info["uses_hess"]
+    use_hessp = use_hessp if use_hessp is not None else method_info["uses_hessp"]
+
+    if use_hess and use_hessp:
+        use_hess = False
+
+    return use_grad, use_hess, use_hessp
+
+
+def get_nearest_psd(A: np.ndarray) -> np.ndarray:
+    """
+    Compute the nearest positive semi-definite matrix to a given matrix.
+
+    This function takes a square matrix and returns the nearest positive semi-definite matrix using
+    eigenvalue decomposition. It ensures all eigenvalues are non-negative. The "nearest" matrix is defined in terms
+    of the Frobenius norm.
+
+    Parameters
+    ----------
+    A : np.ndarray
+        Input square matrix.
+
+    Returns
+    -------
+    np.ndarray
+        The nearest positive semi-definite matrix to the input matrix.
+    """
+    C = (A + A.T) / 2
+    eigval, eigvec = np.linalg.eig(C)
+    eigval[eigval < 0] = 0
+
+    return eigvec @ np.diag(eigval) @ eigvec.T
+
+
+def _unconstrained_vector_to_constrained_rvs(model):
+    constrained_rvs, unconstrained_vector = join_nonshared_inputs(
+        model.initial_point(),
+        inputs=model.value_vars,
+        outputs=get_default_varnames(model.unobserved_value_vars, include_transformed=False),
+    )
+
+    unconstrained_vector.name = "unconstrained_vector"
+    return constrained_rvs, unconstrained_vector
+
+
+def _create_transformed_draws(H_inv, slices, out_shapes, posterior_draws, model, chains, draws):
+    X = pt.tensor("transformed_draws", shape=(chains, draws, H_inv.shape[0]))
+    out = []
+    for rv, idx in slices.items():
+        f = model.rvs_to_transforms[rv]
+        untransformed_X = f.backward(X[..., idx]) if f is not None else X[..., idx]
+
+        if rv in out_shapes:
+            new_shape = (chains, draws) + out_shapes[rv]
+            untransformed_X = untransformed_X.reshape(new_shape)
+
+        out.append(untransformed_X)
+
+    f_untransform = pytensor.function([X], out, mode="JAX")
+    return f_untransform(posterior_draws)
```

Comment:
Do we want to compile a function for a single eval? When I checked in JAX I found it was not worth it to e.g., JIT it. I would check if `mode=Mode(linker="py", optimizer=None)` would be worth it here. Also use `pytensor.In(borrow=True)` and `pytensor.Out(borrow=True)` to avoid deepcopy of variables that are not transformed.

Reply:
yeah that's reasonable. I think the mode='JAX' is just a left-over from an intermediate stage of refactor when I was still hung up on JAX